File: agents/src/vehicle/VehicleAgent.js
// ==============================================
// Vehicle Agent — Fleet capacity & feasibility
// ==============================================

const BaseAgent = require('../base/BaseAgent');
const EVENTS = require('../../../shared/constants/events');

class VehicleAgent extends BaseAgent {
  constructor(eventBus, aiClient) {
    super('vehicle-agent', eventBus, aiClient);
    this.fleet = new Map();
  }

  _listen() {
    // Track vehicle location updates
    this.eventBus.on(EVENTS.VEHICLE_LOCATION_UPDATE, (vehicle) => {
      if (!this.active) return;
      this.fleet.set(vehicle.id, {
        ...this.fleet.get(vehicle.id),
        current_lat: vehicle.current_lat || vehicle.lat,
        current_lng: vehicle.current_lng || vehicle.lng,
        fuel_level: vehicle.fuel_level,
        updated_at: new Date().toISOString(),
      });
    });

    // Handle assignment requests
    this.eventBus.on(EVENTS.ASSIGNMENT_REQUEST, async (request) => {
      if (!this.active) return;
      try {
        const result = await this.findBestVehicle(request.parcels, request.route_distance || 10);
        if (result) {
          this.eventBus.emit(EVENTS.VEHICLE_ASSIGNED, {
            vehicle: result.vehicle,
            shipment_ids: request.parcels.map(p => p.id),
            score: result.score,
          });
        }
        this._recordTask();
      } catch (err) {
        this._recordError(err);
      }
    });

    // Track vehicle availability changes
    this.eventBus.on(EVENTS.VEHICLE_AVAILABLE, (vehicle) => {
      this.fleet.set(vehicle.id, { ...vehicle, status: 'available' });
      this.log.agent(`Vehicle ${vehicle.id} now available`);
    });

    this.eventBus.on(EVENTS.VEHICLE_UNAVAILABLE, (vehicle) => {
      this.fleet.set(vehicle.id, { ...vehicle, status: 'unavailable' });
    });
  }

  /**
   * Register fleet vehicles for tracking.
   */
  registerFleet(vehicles) {
    vehicles.forEach(v => this.fleet.set(v.id, v));
    this.log.agent(`Registered ${vehicles.length} vehicles`);
  }

  /**
   * Check if a vehicle can handle the given parcels and route.
   */
  checkFeasibility(vehicle, parcels, routeDistanceKm) {
    const totalWeight = parcels.reduce((sum, p) => sum + (p.weight_kg || 0), 0);
    const capacity = vehicle.capacity_kg || 200;
    const fuel = vehicle.fuel_level || 0;
    const fuelNeeded = routeDistanceKm * 2; // 2% per km estimate

    const canCarry = totalWeight <= capacity;
    const hasFuel = fuel >= fuelNeeded;
    const isAvailable = vehicle.status === 'available';

    const reasons = [];
    if (!canCarry) reasons.push(`Over capacity: ${totalWeight}/${capacity} kg`);
    if (!hasFuel) reasons.push(`Low fuel: ${fuel}% < ${fuelNeeded}% needed`);
    if (!isAvailable) reasons.push(`Status: ${vehicle.status}`);

    return {
      feasible: canCarry && hasFuel && isAvailable,
      vehicle_id: vehicle.id,
      total_weight_kg: totalWeight,
      capacity_kg: capacity,
      utilization: Math.round((totalWeight / capacity) * 1000) / 10,
      fuel_level: fuel,
      fuel_needed: Math.round(fuelNeeded * 10) / 10,
      reasons,
    };
  }

  /**
   * Find the best vehicle for a set of parcels.
   */
  async findBestVehicle(parcels, routeDistanceKm) {
    const vehicles = Array.from(this.fleet.values());
    const feasible = [];

    for (const v of vehicles) {
      const check = this.checkFeasibility(v, parcels, routeDistanceKm);
      if (check.feasible) {
        const score = 100 - check.utilization * 0.3 - (100 - (v.fuel_level || 100)) * 0.2;
        feasible.push({ vehicle: v, ...check, score: Math.round(score * 10) / 10 });
      }
    }

    feasible.sort((a, b) => b.score - a.score);

    if (feasible.length > 0) {
      this.log.agent(`Best vehicle for ${parcels.length} parcels: ${feasible[0].vehicle_id} (score: ${feasible[0].score})`);
      return feasible[0];
    }

    this.log.warn(`No feasible vehicle found for ${parcels.length} parcels`);
    return null;
  }

  getStatus() {
    const vehicles = Array.from(this.fleet.values());
    return {
      ...super.getStatus(),
      fleet_size: vehicles.length,
      available: vehicles.filter(v => v.status === 'available').length,
      in_transit: vehicles.filter(v => v.status === 'in_transit').length,
    };
  }
}

module.exports = VehicleAgent;

File: ai-engine/agents/vehicle_agent.py
"""Vehicle Agent — Capacity and feasibility checks."""
from typing import Dict, List, Optional

class VehicleAgent:
    def check_feasibility(self, vehicle: Dict, parcels: List[Dict], route_distance: float) -> Dict:
        total_weight = sum(p.get("weight_kg", 0) for p in parcels)
        capacity = vehicle.get("capacity_kg", 200)
        fuel = vehicle.get("fuel_level", 100)
        fuel_needed = route_distance * 2  # 2% fuel per km estimate
        can_carry = total_weight <= capacity
        has_fuel = fuel >= fuel_needed
        return {
            "feasible": can_carry and has_fuel and vehicle.get("status") == "available",
            "vehicle_id": vehicle.get("id"), "total_weight_kg": total_weight,
            "capacity_kg": capacity, "utilization": round(total_weight / capacity * 100, 1),
            "fuel_level": fuel, "fuel_needed": round(fuel_needed, 1),
            "reasons": [] if (can_carry and has_fuel) else
                ([f"Over capacity: {total_weight}/{capacity} kg"] if not can_carry else []) +
                ([f"Low fuel: {fuel}% < {fuel_needed}% needed"] if not has_fuel else [])
        }

    def rank_vehicles(self, vehicles: List[Dict], parcels: List[Dict], route_distance: float) -> List[Dict]:
        results = []
        for v in vehicles:
            check = self.check_feasibility(v, parcels, route_distance)
            if check["feasible"]:
                score = 100 - check["utilization"] * 0.3 - (100 - v.get("fuel_level", 100)) * 0.2
                check["score"] = round(score, 1)
                results.append(check)
        return sorted(results, key=lambda x: x["score"], reverse=True)

_agent = None
def get_vehicle_agent():
    global _agent
    if _agent is None: _agent = VehicleAgent()
    return _agent
